File: leetcode/c++/0011-container-with-most-water.cc
// https://leetcode.com/problems/container-with-most-water/
// Medium

#include<vector>
#include<algorithm>
using namespace std;

class Solution {
public:
    int maxArea(vector<int>& height) {
        int p = 0, maxCount = 0, tempCount = 0, size = height.size();
        int q = size - 1;
        while (p < q) {
            maxCount = max(min(height[p],height[q])*(q-p),maxCount);
            if (height[p] < height[q]) {
                ++p;
            } else {
                --q;
            }
            // Can optimize more by reducing both p and q to the point where
            //      where both their heights are larger than the current min
        }
        return maxCount;
    }
};

File: leetcode/python/0011-container-with-most-water.py
# https://leetcode.com/problems/container-with-most-water/submissions/
# Medium

class Solution:
    def maxArea(self, height: List[int]) -> int:
        lo = 0
        hi = len(height)-1
        m = 0
        while lo < hi:
            h = min(height[lo], height[hi])
            if h*(hi-lo) > m:
                m = h*(hi-lo)
            if height[lo] > height[hi]:
                hi -= 1
            else:
                lo += 1
        return m
        